### PR TITLE
Multicore Cosimulation and Checkpointing

### PR DESCRIPTION
**Pull request by farzamgl**
- Modified the boot code to load in different values for each mhartid
- Added SC to the instructions which are overwritten in cosim
- Fixed a bug with LR which so it clears other cores' reservations

### CODE REVIEW

File: src/dromajo_cosim.cpp [hunk 1]
```diff
@@ -162,6 +171,10 @@ static inline void handle_dut_overrides(RISCVCPUState *s,
           csrno == 0x144 /* sip */)))
         riscv_set_reg(s, rd, dut_wdata);
 
+    /* Catch Store Conditionals */
+    if(is_sc(insn) && rd != 0)
```
Comment (dpetrisko) [inline src/dromajo_cosim.cpp]: super nit: if (

File: src/riscv_cpu.cpp [hunk 2]
```diff
@@ -2162,75 +2171,97 @@ static uint32_t create_fld(int rd, int rs1)
...
     uint32_t data_off = sizeof(uint32_t) * (*data_pos - *code_pos);
 
     rom[(*code_pos)++] = create_auipc(1, data_off);
     rom[(*code_pos)++] = create_addi(1, data_off);
+    rom[(*code_pos)++] = create_add(1, 1, 3);
     rom[(*code_pos)++] = create_ld(1, 1);
 
-    rom[(*data_pos)++] = addr & 0xFFFFFFFF;
-    rom[(*data_pos)++] = addr >> 32;
+    rom[core_offset + (*data_pos)++] = addr & 0xFFFFFFFF;
```
Comment (dpetrisko) [inline src/riscv_cpu.cpp:2249]: So the organization of the rom is:
[core 0 reg 0
 core 0 reg 1
  ...
 core 1 reg 0
 core 1 reg 1
 ...
]
?

File: src/dromajo_cosim.cpp [hunk 3]
```diff
@@ -94,6 +94,15 @@ static inline bool is_amo(uint32_t insn)
     }
 }
 
+static inline bool is_sc(uint32_t insn)
+{
+    int opcode = insn & 0x7f;
+    int func = insn >> 27;
+    if(opcode == 0x2f && func == 3)
```
Comment (dpetrisko) [inline src/dromajo_cosim.cpp]: Super nit: if (

**Review by dpetrisko — APPROVED** [review]
LGTM!

File: src/riscv_cpu.cpp (hunk 2 above)
Comment (farzamgl) [reply src/riscv_cpu.cpp:2249]: Yes.

**Pull request merged**